livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java

```diff
@@ -148,7 +167,23 @@ private String extractStatementResult(String result) {
 
   @Override
   public void cancel(InterpreterContext context) {
-    //TODO(zjffdu). Use livy cancel api which is available in livy 0.3
+    if (livyVersion.isCancelSupported()) {
+      try {
+        if (paragraphId2StmtIdMapping.containsKey(context.getParagraphId())) {
+          cancelStatement(paragraphId2StmtIdMapping.get(context.getParagraphId()));
```

Comment:
instead of
```
if (paragraphId2StmtIdMapping.containsKey(context.getParagraphId())) {
           cancelStatement(paragraphId2StmtIdMapping.get(context.getParagraphId()));
```

should it be
```
id = context.getParagraphId();
if (id != null) {
    stmt = paragraphId2StmtIdMapping.get(id);
    if (stmt != null) {
           cancelStatement(stmt);
```
Since ConcurrentHashMap.get only throws if key is null and returns null if key not found. If we check containsKey first is there a chance for race condition and the ConcurrentHashMap is updated when we call get()?